### ISSUE TITLE
Session hangs after SECC Discovery Protocol Req/Res (TLS )

### ISSUE THREAD

**Issue by zacharyalutz** [Sevenstax/FreeV2G#75]
### Setup
- Platform: **EVSE**
- Firmware Version: **V02_00_01**
- Host Controller Interface: **Ethernet**
- Host: **Own Implementation**
- Payment selection: **EIM only** (no PnC)
- SDP configuration: **default**
- Vehicle: **Ford Lightning**

### Description
Charging session with Ford Lightning progresses normally up to SECC Discovery Protocol request/response message pair. The vehicle requests TLS, and Whitebeet responds that a TLS-secured port is open. No further requests sent by vehicle. 

If you inspect my configuration sequence, I never issue the command for _Set SDP Config_ [```27 68 ...```], so I imagine that I get the default setting. In previous firmware versions, I found that it was not possible to allow secure ports (see #12), and so stopped using _Set SDP Config_ entirely. This configuration, by and large, has been successful but I have not yet come across a vehicle explicitly requesting TLS until now. 

After this failed attempt, the Ford Lightning was tested successfully on a DC fast charger with a different communication platform. (entered main charge loop, no errors).

### To Reproduce
Configure Whitebeet as seen in wireshark log, plug Ford Lightning into EVSE.

### Expected behavior
My expectation depends on the default configuration for SDP. Consider the following scenarios:

- **Secure ports disabled** - Vehicle requests TCP+TLS. Whitebeet responds that only TCP is available. Session continues with Supported Application Protocol Req/Res if plain old TCP is satisfactory to the vehicle. If not, the session ends.
- **Secure ports enabled** - Vehicle requests TCP+TLS. Whitebeet responds that TCP+TLS is available. Session continues with...¯\\\_(ツ)\_/¯

### Remarks
The fact that there is no subsequent request from the vehicle after this point in time, makes me wonder if the response sent by Whitebeet to EV is malformed or otherwise incomplete. Also, it's worth mentioning that I could have goofed up my host implementation w.r.t. V2G serivice configuration, so please let me know if you see any mistakes on my behalf :)

Is it possible that Whitebeet might advertise TLS without actually using it, i.e. secure ports were not enabled by host, but Whitebeet is simply reporting that it supports TLS but does not initiate the TLS handshaking process? I am not entirely sure what the message should look like on the Whitebeet-vehicle side, though I imagine that EVSE presents certificate to EV for authentication, or some such handshaking sequence.  

On another note, I am confused about configuration options regarding secure ports/TLS/PnC. We have one command for SDP port configuration (secure/unsecure/both). In _Set Configuration_ [```27 60```] we have options for certificate support (I imagine these are for TLS functionality), but these options are contingent upon PnC according to the user manual. Certainly, PnC implies TLS (ISO 15118-2 requires it), but TLS does not necessarily imply PnC. It seems that Whitebeet imposes this erroneous relationship: TLS<=>PnC.  Can you specify use cases for the combinations of all settings (SDP port configuration/certificate installation support/certificate update support/payment method). What combinations are meaningful, and what combinations are un-meaningful? A factorial table including all combinations would be extremely helpful.
 
#### TL;DR -
Consider what Whitebeet EVSE should do if vehicle requests TCP+TLS knowing that PnC is not a payment option. 
  
### Logs
[ford-lightning-7-13-SDP-fail.pcapng.gz](https://github.com/Sevenstax/FreeV2G/files/9104941/ford-lightning-7-13-SDP-fail.pcapng.gz)

**Comment by jpo-stx** [comment]
Hi @zacharyalutz ,

> Charging session with Ford Lightning progresses normally up to SECC Discovery Protocol request/response message pair. The vehicle requests TLS, and Whitebeet responds that a TLS-secured port is open. No further requests sent by vehicle.

This is really strange, if the vehicle requests a secure connection and the Whitebeet offeres one we should see a TCP SYN from the EV on the port sent in the SDP response message. Port 56466 in your case.

> After this failed attempt, the Ford Lightning was tested successfully on a DC fast charger with a different communication platform. (entered main charge loop, no errors).

Did the other platform also offered a secure connection in the SDP response? It is the second to last byte in the SDP response. `0x00` is secured and `0x01` is unsecured.

> The fact that there is no subsequent request from the vehicle after this point in time, makes me wonder if the response sent by Whitebeet to EV is malformed or otherwise incomplete. Also, it's worth mentioning that I could have goofed up my host implementation w.r.t. V2G serivice configuration, so please let me know if you see any mistakes on my behalf :)

Your `Set Configuration` and `Set DC Charging Parameters` messages are OK.

> Is it possible that Whitebeet might advertise TLS without actually using it, i.e. secure ports were not enabled by host, but Whitebeet is simply reporting that it supports TLS but does not initiate the TLS handshaking process? I am not entirely sure what the message should look like on the Whitebeet-vehicle side, though I imagine that EVSE presents certificate to EV for authentication, or some such handshaking sequence.

There might me some confusion and the manual is not clearly defining the use cases. What you want to do is only support unsecrue connections with EIM and Whitebeet is offering a secure connection if the EV requests it, despite not being able to open a secure connection (no TLS certificates set I assume). In this case a secure connection shouldn't be offered. This needs to be improved.

> On another note, I am confused about configuration options regarding secure ports/TLS/PnC. We have one command for SDP port configuration (secure/unsecure/both). In Set Configuration [27 60] we have options for certificate support (I imagine these are for TLS functionality), but these options are contingent upon PnC according to the user manual. Certainly, PnC implies TLS (ISO 15118-2 requires it), but TLS does not necessarily imply PnC. It seems that Whitebeet imposes this erroneous relationship: TLS<=>PnC. Can you specify use cases for the combinations of all settings (SDP port configuration/certificate installation support/certificate update support/payment method). What combinations are meaningful, and what combinations are un-meaningful? A factorial table including all combinations would be extremely helpful.

The certificate support is only needed for PNC charging. With these options you can offer the installation or updating of the contract certificate of the EV which is used for the payment. To support this you need a connection to a backend which supports this functionality.

Yes, we need to improve the description of the usecases. There are 3 possibilities:
1. EIM with an unsecure connection
2. EIM with a secure connention
3. PNC with a secure connection

I think the main problem is that the Whitebeet is offering a secure connection without having the needed certificates. In the default configuration it should only offer a secure connection when the TLS certificates are available.

### Workaround
As a workaround you can try to set the SDP config from the example. This disables the secure connection.
